llvm-bazel/llvm-project-overlay/mlir/tblgen.bzl

```diff
@@ -4,7 +4,128 @@
 
 """BUILD extensions for MLIR table generation."""
 
-def gentbl(name, tblgen, td_file, tbl_outs, td_srcs = [], td_includes = [], td_relative_includes = [], strip_include_prefix = None, test = False):
+# A provider with one field, transitive_sources.
+TdFiles = provider(fields = ["transitive_sources", "transitive_includes"])
+
+def _get_dep_transitive_srcs(dep):
+    if TdFiles in dep:
+        return dep[TdFiles].transitive_sources
+    return dep[DefaultInfo].files
+
+def _get_dep_transitive_includes(dep):
+    if TdFiles in dep:
+        return dep[TdFiles].transitive_includes
+    return depset()
+
+def _get_transitive_srcs(srcs, deps):
+    """Obtain the source files for a target and its transitive dependencies.
+
+    Args:
+      srcs: a list of source files
+      deps: a list of targets that are direct dependencies
+    Returns:
+      a collection of the transitive sources
+    """
+    return depset(
+        srcs,
+        transitive = [_get_dep_transitive_srcs(dep) for dep in deps],
+    )
+
+def _get_transitive_includes(includes, deps):
+    return depset(
+        includes,
+        transitive = [_get_dep_transitive_includes(dep) for dep in deps],
+    )
+
+def _td_library_impl(ctx):
+    trans_srcs = _get_transitive_srcs(ctx.files.srcs, ctx.attr.deps)
+    trans_includes = _get_transitive_includes(ctx.attr.includes, ctx.attr.deps)
+    return [TdFiles(
+        transitive_sources = trans_srcs,
+        transitive_includes = trans_includes,
+    )]
+
+td_library = rule(
+    _td_library_impl,
+    attrs = {
+        "srcs": attr.label_list(allow_files = True),
+        "includes": attr.string_list(),
+        "deps": attr.label_list(),
```

Comment:
I think it makes sense to do that for gentbl_rule regardless (which I should probably rename, actually) since deps was not a pre-existing attribute. For this one, see my comment above.

Reply:
That makes good sense!